### ISSUE TITLE
Transaction peer recovery

### ISSUE THREAD

**Issue by atosak** [OpenLiberty/open-liberty#8968]
This feature provides autonomous transactional peer recovery among groups of peer servers either in a cloud or otherwise.
This will allow 2 phase commit to be used in cloud environments for the first time and will enable the clearing of database locks held by servers that have been "reaped" by cloud runtimes.

**Comment by atosak** [comment]
Resolves RFE 190360[](https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/External%20Request%20For%20Enhancements#action=com.ibm.team.workitem.viewWorkItem&id=190360)
Resolves RFE [174417[](https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/External%20Request%20For%20Enhancements#action=com.ibm.team.workitem.viewWorkItem&id=174417)]

**Comment by irobins** [comment]
Jon - is there a draft written design, where can I find it and how soon are we hoping to schedule a review?
As we discussed this week, I'm bothered about the details of practicality and testing of the persistent store. Details like:

- In OpenShift, how is the transaction log persistence configured, and what are the volume or DB prereqs, to ensure common access to the log across the scope of the recovery group?
- What are the deployment limitations - for example, if using a persistent volume, what configuration is required to ensure the PV is mounted in a way that ensure there is always one writer. I'd assume an access mode of ReadWriteMany (which constrains the volume filesystem to NFS per https://docs.openshift.com/container-platform/4.2/storage/understanding-persistent-storage.html). Would there then need to be a dependency on NFSv4 locking lease (as used in WAS HA tx logging today)? Do we have that code in Liberty? 
- What does the test scenario look like for this and when can we plan to validate it?
- Can we also plan to get some performance data on an OpenShift environment for throughput any options (e.g. persistent volume vs database) related to 
-

**Comment by jonhawkes** [comment]
UFO: https://ibm.box.com/s/cki7qez2eiqkm91hr3ifng6v3n8boti8

**Comment by jonhawkes** [comment]
Feature Test Summary: https://github.com/OpenLiberty/open-liberty/issues/16112

**Comment by jonhawkes** [comment]
## Description of the high level feature, including any external spec links:  
This feature provides autonomous transactional peer recovery among groups of peer servers either in a cloud or otherwise.
This will allow 2 phase commit to be used in cloud environments for the first time and will enable the clearing of database locks held by servers that have been "reaped" by cloud runtimes.
<br/>
##
#### Before proceeding to any items below (active development), this feature must be prioritized on the backlog, and have either the "Design Approved" or "No Design Approved" labels.  Follow the Feature and UFO Approval Process.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/cki7qez2eiqkm91hr3ifng6v3n8boti8
- FTS: https://github.com/OpenLiberty/open-liberty/issues/16112
- Beta Blog Post (if applicable): #18328 
- Blog Post: 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [ ] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [ ] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.


## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [x] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [x] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by tjwatson** [comment]
UFO - Feedback/Actions:

- Communication
  - must clearly document restrictions and constraints for file system persistence
- To-Be
  - Need to determine documentation required for open-liberty vs knowledge center.
  - SVT must demonstrate support in OpenShift
  - DDL generation should be done through the persistence service, currently hard coded for each supported DB.
- Feature Design
  - Keep leaseCheckStrategy as an internal.  Will allow it to be public if/when we support more than just round-robin
  - Liberty operator should be enhanced to automatically set the recoveryGroup with an environment variable. Default config should point to the environment variable to allow easy configuration in cloud environment
  - Name leaseRenewalTime is confusing.  Need suggestions for alternate name - LeaseRenewalThreshold?
  - Should leaseLength/leaseCheckInterval be one setting?  (not really sure if there was consensus on this one)
- Admin / Config / Command line
  - Be sure to document multi-zone environments not supported for file system based.
- Beta
  - We can beta this in the next beta release if appropriate
- System Test Impact
  - Be sure to call out OpenShift testing
- Performance
  - Discussion on 90% lease renewal.  Would checking twice before expiration make for a better experience?
- Migration
  - Need to review migration rules around support for 2 phase commit (Cindy High action?)
  - What can be done to help migration the support in tWAS to the configuration in Liberty.  Does not appear to map well, but we do need to at least guide the user to the proper documentation in Liberty to do what is needed

**Comment by jonhawkes** [comment]
#16362 raised to address DDL generation issue.

**Comment by NottyCode** [comment]
@jonhawkes can you let me know once the actions from the review have been resolved?

**Comment by skasund** [comment]
@jonhawkes   L2 has requested a live STE presentation for this feature.  The STE templates can be found at the links below.  You can use either one to create the education.  

Slide Template:  https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Github Template:  https://pages.github.ibm.com/WASL3/site/STE/about

Please upload the completed slides to the same 'STE Archive' BOX folder or provide me the Github link, and I will provide the STE approval.  The live presentation can be scheduled with the L2 team at a later date.  Thanks!

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  **Not explicitly but there were no serviceability questions raised during the UFO review.** Have these issues been addressed in the implementation? **Unusually, the code has been shipped and serviced in the Liberty transactions component for several years.**

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? **Server failure and mis-configuration**
b) Who did you demo to? **WAS L3: Trans., Messaging, OSGI, Cscope. Mission Team**
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3? **Yes**

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature? **Tam Dinh**
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  **Yes**

4.  Which L2 / L3 queues will handle PMRs for this feature?  [WASCET/WL3BKT](http://l3tools1.fyre.ibm.com:9080/tororeport/QueueContactSummary.jsp#WL3BKT) Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info. **See STE documentation: https://ibm.box.com/s/flof2hvqn4yupqvv2bx7ka8hymzk1nev**

5.  Does this feature add any new metrics or emit any new JSON events?  **No.**  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by samwatibm** [comment]
Implements https://cloud-platform.ideas.ibm.com/ideas/TWAS-I-123

**Comment by chirp1** [comment]
John approved David's draft in ID issue https://github.com/OpenLiberty/docs/issues/4423. Approving the epic.
